crates/shared/src/price_estimation/buffered.rs

```diff
@@ -0,0 +1,577 @@
+        num::NonZeroUsize,
+        sync::Arc,
+        time::Duration,
+    },
+    tokio::{sync::broadcast, task::JoinHandle, time::error::Elapsed},
+};
+
+/// Buffered configuration.
+#[derive(Clone)]
+#[allow(dead_code)]
+pub struct Configuration {
+    /// The maximum amount of concurrent batches to request.
+    ///
+    /// Specifying `None` means no limit on concurrency.
+    pub max_concurrent_requests: Option<NonZeroUsize>,
+    /// The maximum batch size.
+    pub max_batch_len: usize,
+    /// An additional minimum delay to wait for collecting requests.
+    ///
+    /// The delay starts counting after receiving the first request.
+    pub batch_delay: Duration,
+    /// The timeout to wait for the result to be ready
+    pub result_ready_timeout: Duration,
+    /// Maximum capacity of the broadcast channel to store the native prices
+    /// results
+    pub broadcast_channel_capacity: usize,
+}
+
+/// Trait for fetching a batch of native price estimates.
+#[allow(dead_code)]
+#[cfg_attr(test, mockall::automock)]
+pub trait NativePriceBatchFetching: Sync + Send + NativePriceEstimating {
+    /// Fetches a batch of native price estimates.
+    ///
+    /// It returns a HashMap which maps the token with its native price
+    /// estimator result
+    fn fetch_native_prices(
+        &self,
+        tokens: &HashSet<H160>,
+    ) -> futures::future::BoxFuture<
+        '_,
+        Result<HashMap<H160, NativePriceEstimateResult>, PriceEstimationError>,
+    >;
+}
+
+/// Buffered implementation that implements automatic batching of
+/// native prices requests.
+#[allow(dead_code)]
+#[derive(Clone)]
+pub struct BufferedRequest<Inner> {
+    config: Configuration,
+    inner: Arc<Inner>,
+    calls: mpsc::UnboundedSender<H160>,
+    broadcast_sender: broadcast::Sender<NativePriceResult>,
+}
+
+/// Object to map the token with its native price estimator result
+#[allow(dead_code)]
+#[derive(Clone)]
+struct NativePriceResult {
+    token: H160,
+    result: Result<f64, PriceEstimationError>,
+}
+
+impl<Inner> NativePriceEstimating for BufferedRequest<Inner>
+where
+    Inner: NativePriceBatchFetching + NativePriceEstimating + 'static,
+{
+    /// Request to get estimate prices in a batch
+    fn estimate_native_price(
+        &self,
+        token: H160,
+    ) -> futures::future::BoxFuture<'_, NativePriceEstimateResult> {
+        async move {
+            // Sends the token for requesting price
+            self.calls.unbounded_send(token).map_err(|e| {
+                PriceEstimationError::ProtocolInternal(anyhow!(
+                    "failed to append a new token to the queue: {e:?}"
+                ))
+            })?;
+
+            let mut rx = self.broadcast_sender.subscribe();
+
+            tokio::time::timeout(self.config.result_ready_timeout, async {
+                loop {
+                    if let Ok(Some(result)) =
+                        Self::receive_with_timeout(&mut rx, &token, self.config.batch_delay).await
```

Comment:
The double timeout seems confusing and wrong to me. I'm ok with a timeout (although I generally think we are making the config overly complicated) but two timeouts don't make sense.

Reply:
I agree that two timeout is a little bit overkill, but I wanted to be as defensive as possible. Changing it to one timeout :ok_hand:

Reply:
btw @fleupold , removing the double timeout doesn't reduce the configuration fields. But I will add default values :ok_hand:

Reply:
ummm I didn't propagate the config to the autopilot/orderbook config yet (I plant to do that in the 3/3 PR). Maybe there it's where we should have the defaults? (alongside the other default config values) :thinking: